File: Sources/TokenTrackerMenuBar/DiagnosticsReporter.swift
import Foundation
import TokenTrackerCore

@MainActor
struct DiagnosticsReporter {
    let settings: Settings
    let historyStore: UsageHistoryStore
    let snapshot: UsageSnapshot?
    let lastSuccessfulRefreshAt: Date?
    let runningInstanceCount: Int

    func diagnosticsText() -> String {
        var lines: [String] = []
        lines.append("Token Tracker Diagnostics")
        lines.append("Generated: \(isoString(Date()))")
        lines.append("App version: \(appVersion) (\(appBuild))")
        lines.append("Bundle id: \(Bundle.main.bundleIdentifier ?? "unknown")")
        lines.append("macOS: \(ProcessInfo.processInfo.operatingSystemVersionString)")
        lines.append("Architecture: \(machineArchitecture())")
        lines.append("Display mode: \(settings.displayMode.rawValue)")
        lines.append("Provider labels: \(settings.providerLabelStyle.rawValue)")
        lines.append("Refresh interval: \(Int(settings.refreshInterval))s")
        lines.append("Claude enabled: \(settings.claudeEnabled)")
        lines.append("Codex enabled: \(settings.codexEnabled)")
        lines.append("Language: \(settings.language.rawValue)")
        lines.append("Notifications enabled: \(settings.notificationsEnabled)")
        lines.append("5h alert threshold: \(settings.fiveHourAlertThreshold)%")
        lines.append("7d alert threshold: \(settings.sevenDayAlertThreshold)%")
        lines.append("Reset alert window: \(settings.resetAlertMinutes)m")
        lines.append("History retention: \(settings.historyRetentionDays)d")
        lines.append("History entries: \(historyStore.load().count)")
        lines.append("History trend: \(historyTrendText(language: .english))")
        lines.append("Last successful update: \(lastSuccessfulRefreshAt.map(isoString) ?? "none")")
        lines.append("Running instances: \(runningInstanceCount)")
        if settings.refreshInterval < 60 {
            lines.append("Refresh warning: \(Localizer(language: .english).text(.refreshIntervalWarning))")
        }
        // Not just "does the fallback file exist" — on a normal macOS install it does
        // not, and the answer that matters is which store is live and whether anything
        // is still writing to it. The Claude access token lives ~8h and only the CLI
        // refreshes this store, so an age well past that is the whole explanation for
        // a token that keeps lapsing.
        let claudeSource = ClaudeCredentialSource.detect(credentialsFileURL: Self.claudeCredentialsURL)
        lines.append("Claude credential source: \(claudeSource.kind.rawValue)")
        lines.append("Claude credential last written: \(claudeSource.lastWrittenAt.map(isoString) ?? "unknown")")
        lines.append("Claude credential age: \(claudeSource.age().map(Self.formatAge) ?? "unknown")")
        lines.append("Claude credentials file exists: \(fileExists(at: Self.claudeCredentialsURL))")
        lines.append("Codex auth file exists: \(fileExists(at: Self.codexAuthURL))")

        if let snapshot {
            lines.append("Snapshot updated: \(isoString(snapshot.updatedAt))")
            lines.append(contentsOf: diagnosticsLines(for: snapshot.claude))
            lines.append(contentsOf: diagnosticsLines(for: snapshot.codex))
        } else {
            lines.append("Snapshot: none")
        }

        return lines.joined(separator: "\n")
    }

    func historyTrendText(language: AppLanguage? = nil) -> String {
        guard let snapshot else {
            return Localizer(language: language ?? settings.language).text(.notEnoughHistory)
        }
        return UsageHistoryFormatter.trendSummary(
            entries: historyStore.load(),
            current: snapshot,
            localizer: Localizer(language: language ?? settings.language)
        )
    }

    static var claudeCredentialsURL: URL {
        FileManager.default.homeDirectoryForCurrentUser.appendingPathComponent(".claude/.credentials.json")
    }

    static var codexAuthURL: URL {
        FileManager.default.homeDirectoryForCurrentUser.appendingPathComponent(".codex/auth.json")
    }

    private func diagnosticsLines(for usage: ProviderUsage) -> [String] {
        let issue = UsageIssueFormatter.issue(for: usage, localizer: Localizer(language: .english))
        return [
            "\(usage.provider.displayName) source: \(usage.source.rawValue)",
            "\(usage.provider.displayName) status: \(issue.kind.rawValue)",
            "\(usage.provider.displayName) 5h remaining: \(DisplayFormatter.formatPercent(usage.remainingPercent5h))",
            "\(usage.provider.displayName) 7d remaining: \(DisplayFormatter.formatPercent(usage.remainingPercent7d))",
            "\(usage.provider.displayName) 5h reset: \(isoStringOrDash(usage.resetAt5h))",
            "\(usage.provider.displayName) 7d reset: \(isoStringOrDash(usage.resetAt7d))",
            "\(usage.provider.displayName) plan: \(usage.plan ?? "--")",
            "\(usage.provider.displayName) technical error: \(issue.technicalDetail ?? "--")"
        ]
    }

    private var appVersion: String {
        Bundle.main.object(forInfoDictionaryKey: "CFBundleShortVersionString") as? String ?? "unknown"
    }

    private var appBuild: String {
        Bundle.main.object(forInfoDictionaryKey: "CFBundleVersion") as? String ?? "unknown"
    }

    private func machineArchitecture() -> String {
        #if arch(arm64)
        return "arm64"
        #elseif arch(x86_64)
        return "x86_64"
        #else
        return "unknown"
        #endif
    }

    private func fileExists(at url: URL) -> Bool {
        FileManager.default.fileExists(atPath: url.path)
    }

    private static func formatAge(_ age: TimeInterval) -> String {
        let minutes = max(0, Int(age) / 60)
        if minutes < 60 { return "\(minutes)m" }
        let hours = minutes / 60
        if hours < 24 { return "\(hours)h \(minutes % 60)m" }
        return "\(hours / 24)d \(hours % 24)h"
    }

    private func isoString(_ date: Date) -> String {
        ISO8601DateFormatter().string(from: date)
    }

    private func isoStringOrDash(_ date: Date?) -> String {
        guard let date else { return "--" }
        return isoString(date)
    }
}

File: windows/TokenTracker.Windows/DiagnosticsReporter.cs
using System.Reflection;
using System.Runtime.InteropServices;
using TokenTracker.Windows.Core;

namespace TokenTracker.Windows;

internal sealed class DiagnosticsReporter
{
    private readonly AppSettings settings;
    private readonly UsageHistoryStore historyStore;
    private readonly UsageSnapshot? snapshot;
    private readonly DateTimeOffset? lastSuccessfulRefreshAt;
    private readonly int runningInstanceCount;

    public DiagnosticsReporter(
        AppSettings settings,
        UsageHistoryStore historyStore,
        UsageSnapshot? snapshot,
        DateTimeOffset? lastSuccessfulRefreshAt,
        int runningInstanceCount)
    {
        this.settings = settings;
        this.historyStore = historyStore;
        this.snapshot = snapshot;
        this.lastSuccessfulRefreshAt = lastSuccessfulRefreshAt;
        this.runningInstanceCount = runningInstanceCount;
    }

    public string DiagnosticsText()
    {
        var lines = new List<string>
        {
            "Token Tracker Diagnostics",
            $"Generated: {IsoString(DateTimeOffset.Now)}",
            $"App version: {AppVersion}",
            $"OS: {RuntimeInformation.OSDescription}",
            $"Architecture: {RuntimeInformation.ProcessArchitecture}",
            $"Display mode: {settings.DisplayMode}",
            $"Provider labels: {settings.ProviderLabelStyle}",
            $"Refresh interval: {settings.RefreshIntervalSeconds}s",
            $"Claude enabled: {settings.ClaudeEnabled}",
            $"Codex enabled: {settings.CodexEnabled}",
            $"Language: {settings.Language}",
            $"Notifications enabled: {settings.NotificationsEnabled}",
            $"5h alert threshold: {settings.FiveHourAlertThreshold}%",
            $"7d alert threshold: {settings.SevenDayAlertThreshold}%",
            $"Reset alert window: {settings.ResetAlertMinutes}m",
            $"History retention: {settings.HistoryRetentionDays}d",
            $"History entries: {historyStore.Load().Count}",
            $"History trend: {HistoryTrendText(AppLanguage.English)}",
            $"Last successful update: {(lastSuccessfulRefreshAt is null ? "none" : IsoString(lastSuccessfulRefreshAt.Value))}",
            $"Running instances: {runningInstanceCount}"
        };

        if (settings.RefreshIntervalSeconds < 60)
        {
            lines.Add($"Refresh warning: {new Localizer(AppLanguage.English).Text(L10nKey.RefreshIntervalWarning)}");
        }

        // Not just whether the file is there: when usage stops loading, the question
        // that matters is whether anything is still writing to it. The Claude access
        // token lives ~8h, so an age well past that says the store has been abandoned
        // rather than that the token was merely unlucky. The source is reported for
        // parity with macOS, where the keychain usually wins; Windows has no keychain,
        // so the file is always the live store here.
        var claudeSource = ClaudeCredentialSource.Detect(ClaudeCredentialsPath);
        var credentialLastWritten = claudeSource.LastWrittenAt is { } written ? IsoString(written) : "unknown";
        var credentialAge = claudeSource.Age(DateTimeOffset.Now) is { } age
            ? UsageForecaster.DurationText(age.TotalSeconds)
            : "unknown";
        lines.Add($"Claude credential source: {claudeSource.Kind}");
        lines.Add($"Claude credential last written: {credentialLastWritten}");
        lines.Add($"Claude credential age: {credentialAge}");
        lines.Add($"Claude credentials file exists: {File.Exists(ClaudeCredentialsPath)}");
        lines.Add($"Codex auth file exists: {File.Exists(CodexAuthPath)}");

        if (snapshot is null)
        {
            lines.Add("Snapshot: none");
        }
        else
        {
            lines.Add($"Snapshot updated: {IsoString(snapshot.UpdatedAt)}");
            lines.AddRange(DiagnosticsLines(snapshot.Claude));
            lines.AddRange(DiagnosticsLines(snapshot.Codex));
        }

        return string.Join(Environment.NewLine, lines);
    }

    public string HistoryTrendText(AppLanguage? language = null)
    {
        if (snapshot is null)
        {
            return new Localizer(language ?? settings.Language).Text(L10nKey.NotEnoughHistory);
        }

        return UsageHistoryFormatter.TrendSummary(
            historyStore.Load(),
            snapshot,
            localizer: new Localizer(language ?? settings.Language));
    }

    public static string ClaudeCredentialsPath => AppPaths.ClaudeCredentialsPath();

    public static string CodexAuthPath => AppPaths.CodexAuthPath();

    private static IEnumerable<string> DiagnosticsLines(ProviderUsage usage)
    {
        var issue = UsageIssueFormatter.Issue(usage, new Localizer(AppLanguage.English));
        return new[]
        {
            $"{usage.DisplayName} source: {usage.Source}",
            $"{usage.DisplayName} status: {issue.Kind}",
            $"{usage.DisplayName} 5h remaining: {DisplayFormatter.FormatPercent(usage.RemainingPercent5h)}",
            $"{usage.DisplayName} 7d remaining: {DisplayFormatter.FormatPercent(usage.RemainingPercent7d)}",
            $"{usage.DisplayName} 5h reset: {IsoStringOrDash(usage.ResetAt5h)}",
            $"{usage.DisplayName} 7d reset: {IsoStringOrDash(usage.ResetAt7d)}",
            $"{usage.DisplayName} plan: {usage.Plan ?? "--"}",
            $"{usage.DisplayName} technical error: {issue.TechnicalDetail ?? "--"}"
        };
    }

    private static string AppVersion =>
        AppVersionText.Format(Assembly.GetExecutingAssembly().GetName().Version);

    private static string IsoString(DateTimeOffset date) =>
        date.ToUniversalTime().ToString("O");

    private static string IsoStringOrDash(DateTimeOffset? date) =>
        date is null ? "--" : IsoString(date.Value);
}
